### ISSUE TITLE
Hydra with timeout

### ISSUE THREAD

**Issue by PythonNut** [abo-abo/hydra#34]
I'm currently using `smartrep` to implement key-chords in areas where key-chords are unavailable. I'm trying to migrate to `hydra` instead.

Take, for example, the following code, which mimics a <kbd>jj</kbd> ⇒ `escape from isearch` chord.

```emacs
(defun isearch-exit-chord-worker ()
  "Exit out of isearch after a chord"
  (interactive)
  (execute-kbd-macro (kbd "<backspace> <return>")))

(defhydra isearch-exit-hydra
  (:pre (setq hydra-is-helpful nil)
    :post (setq hydra-is-helpful t))
  ("j" isearch-exit-chord-worker)
  ("k" isearch-exit-chord-worker))

(defun isearch-exit-chord (arg)
  (interactive "p")
  (isearch-printing-char)
  (isearch-exit-hydra/body))

(define-key isearch-mode-map "j" #'isearch-exit-chord)
(define-key isearch-mode-map "k" #'isearch-exit-chord)
```
The only problem is that it is impossible to actually type `jj`. I'm my previous solution, I could break out after a certain amount of time has passed (so you can type <kbd>j</kbd> wait... <kbd>j</kbd> to literally insert), but I haven't figured out how to do that in hydras.

**Comment by abo-abo** [comment]
I don't really see what advantage you're getting from hydras here.

Here's a simple approach:

```cl
(defun isearch-exit-chord-worker ()
  "Exit out of isearch after a chord"
  (interactive)
  (if (string-match "j$" isearch-string)
      (progn
        (isearch-delete-char)
        (isearch-done))
    (isearch-printing-char)))

(define-key isearch-mode-map "j" #'isearch-exit-chord-worker)
```

**Comment by PythonNut** [comment]
Yes, but that will make it impossible to type `jj`. You _could_ extend it with timers and such, but using `smartrep` seemed cleaner.

**Comment by abo-abo** [comment]
So you want the action only after not too much time has passed, otherwise insert another <kbd>j</kbd>?

**Comment by PythonNut** [comment]
Yes.

**Comment by abo-abo** [comment]
Would this work:

```cl
(defvar hydra-last-time (current-time))
(defun isearch-exit-chord-worker ()
  "Exit out of isearch after a chord"
  (interactive)
  (let ((td (time-to-seconds
             (time-subtract (current-time)
                            hydra-last-time))))
    (setq hydra-last-time (current-time))
    (if (and (< td 0.5)
             (string-match "j$" isearch-string))
        (progn
          (isearch-delete-char)
          (isearch-done))
      (isearch-printing-char))))
(define-key isearch-mode-map "j" #'isearch-exit-chord-worker)
```

**Comment by PythonNut** [comment]
Thank you. I'm afraid it sill has other problems, but you needn't work on them. I can fix them myself.

Really, this is about adding timeouts to `hydra`. If you don't think it's worth the extra complexity, I'll just go back to using my old solution. I'm fine either way.

**Comment by abo-abo** [comment]
> Thank you. I'm afraid it sill has other problems

Do share, I'm still not clear on what you want to achieve.

**Comment by PythonNut** [comment]
> Do share

* what happens when I'm not at the end of the string? 
* why does <kbd>jz</kbd> <kbd>DEL</kbd> <kbd>j</kbd> trigger as well?

> I'm still not clear on what you want to achieve.
I just want to be able to mash <kbd>jj</kbd> and exit out of isearch.

Please remember, though, that I already have a working solution. Your time is valuable, and you don't need to spend it coming up with another solution. The reason this issue exists is that I believe this would be a useful extension of `hydra` (for other purposes too).

**Comment by abo-abo** [comment]
> what happens when I'm not at the end of the string?

How can that happen? You're always at the end of the string with isearch.

> why does jz DEL j trigger as well?

If you manage to type that in 0.5s that's phenomenal.

**Comment by PythonNut** [comment]
> How can that happen? You're always at the end of the string with isearch.

I want to use this for chords in general. Isearch was my one example use-case among many.

**Comment by abo-abo** [comment]
Is there an update on this? Something that needs to be done on my end?
Otherwise, close the issue.

**Comment by PythonNut** [comment]
@abo-abo 

> Really, this is about adding timeouts to hydra. If you don't think it's worth the extra complexity, I'll just go back to using my old solution. I'm fine either way.

I'm proposing this as a potential enhancement. Each head would have a `:lifetime` property (or whatever clever thing you come up with), after which it would die.

If you think that this would not enhance `hydra` then by all means close this.

**Comment by glucas** [comment]
Interesting idea. There might be cases where right now I use a nil head head to quit out of a hydra that would be cleaner if the hydra simply died on its own after a moment.

**Comment by abo-abo** [comment]
I haven't yet figured out why the timers run indefinitely unless
canceled.

For now, applying `:timeout 2.0` to body will cancel the hydra when no
head has been called in 2 seconds.  Calling a head refreshes the
period.  Eventually, each head should be able to override or cancel
the timer.

**Comment by zhaojiangbin** [comment]
Is there an example of how to use this new feature? I tried adding ```:timeout 2.0``` as a body property. I always got "Wrong type argument: number-or-marker-p, nil" in the echo area.

**Comment by abo-abo** [comment]
@zhaojiangbin, here's an example:

```cl
(defhydra hydra-window (:timeout 2.0)
   "window"
   ("h" windmove-left)
   ("j" windmove-down)
   ("k" windmove-up)
   ("l" windmove-right)
   ("q" nil "cancel"))
(global-set-key (kbd "C-M-o") 'hydra-window/body)
```

As you type <kbd>C-M-o</kbd>, you can wait as long as you want.  But as soon as you press one of the
directions, a two second timer to cancel the Hydra starts.  You reset the timer each time you press
a head. Make sure that you have Emacs 24.4.1, just to avoid possible compatibility issues.

**Comment by glucas** [comment]
@zhaojiangbin I was getting the wrong type argument error as well, and submitted a pull request that has been merged. (Thanks @abo-abo!)

The default behavior is to start the timer once a head is invoked. If you want to start the timer immediately you can call `(hydra-timeout)` in a `:body-pre` hook.

**Comment by zhaojiangbin** [comment]
@glucas The default behavior works for me. After upgrading to the latest MELPA version, I have it working. Thanks.

However timeout does not seem to be calling the :post handler. In the following hydra, the message "test b quit" does not show up if timed out.

It would be nicer if timeout can hide the *LV* window, too. That will give a visual clue about what has happened.

```elisp
(bind-key
 "C-c t"
 (defhydra test-hydra-b
   (:timeout 2.0 :post (message "test b quit"))
   "test b"
   ("x" (message "x"))
   ("y" (message "y"))
   ("q" nil "quit")))
```

**Comment by abo-abo** [comment]
> However timeout does not seem to be calling the :post handler

This is the same problem as in #67. I'll get to it soon.

**Comment by abo-abo** [comment]
The most recent thing should be fixed now.
If something else is needed, just start a new issue, this one is too old.

**Comment by sandric** [comment]
I guess my question may be related to issue, so I post it here. I wonder if its possible to set function on timed out event? I wan't to implement simple behavior - on pressing key binding once - one function is called, if the same keybinding pressed twice with small interval - another. Here's simple setup:
```
(defhydra hydra-test (:timeout 0.2 :post (message "one"))
            "test"
            ("C-H-w" (message "two")))

(global-set-key (kbd "C-H-w") 'hydra-test/body)
```
Here I used post handler - but it calls always - even if timeout passed - as it was fixed upper. What I seek is like exit (I tried it too but looks like its not triggered), cancel or quit, or best of all - "timedout" handler.

Is it doable somehow for now?

**Comment by abo-abo** [comment]
@sandric Just open a new issue describing what you actually want to do.  Preferably something more practical than displaying messages: it's hard to figure out "how" if you don't know "why".

The `:timeout` key should be used for timeout only. If you look at `hydra--make-defun` the `:timeout` is used right after the more generic `:after-exit` key. You can put arbitrary constructs including your own timers into `:after-exit`.

**Comment by sandric** [comment]
Well, I just wanted to make it clear, not bothering with my particular workflow. Ok, maybe it wasn't very clear what I wanted to do - I want to be able to make following:

- Press key combination
- If during 0.2 seconds I didn't repeat pressing that combination - run function A
- If during 0.2 seconds I do repeat the same combination - run function B instead of A

My particular scenario why I want to do that - I want to run "fzf" function if I pressed combination once and run "fzf-directory" (the same fzf but with selecting directory first) if I pressed the same combination twice but quick, here's code
```elisp
(defhydra hydra-fzf (:timeout 0.2 :post 'fzf)
            "fzf"
            ("C-H-w" fzf-directory))

(global-set-key (kbd "C-H-w") 'hydra-test/fzf)
```

**Comment by glucas** [comment]
@sandric The `key-chord` library allows you to create bindings for a repeated (double tapped) key that are independent of any binding that key might have on its own. I think this would do exactly what you want here.

**Comment by sandric** [comment]
@glucas No, you can not, because of this:
```
;; Emacs will not call input-method-function for keys that have non numeric
;; codes or whos code is outside the range 32..126. Thus you cannot define
;; key chords involving function keys, control keys, or even your non-english
;; letters (on national keyboards) that otherwise are well positioned for
;; chording on your keyboard.
;; (I think chording left and right arrow keys would be useful, but cannot do.
;; I consider this a bug in Emacs. Input methods could happily return 
;; unmodified *any* key they don't know about.)
```
And this not eliminate all the other workflows - what if one day you will want to add third function, for example:

```elisp
(defhydra hydra-fzf (:timeout 0.2 :post 'fzf)
            "fzf"
            ("C-H-w" fzf-directory)
            ("C-H-k" fzf-kill-process))

(global-set-key (kbd "C-H-w") 'hydra-fzf/body)
```
Hydra is great in what it stands for (for me) - grouping lots of heads under one body - which is then can be binded to one key combination. All I want is to have place where I can set function to execute when none of those heads selected and timeout is gone. The "default head" so to speak ))

**Comment by abo-abo** [comment]
@sandric 

Try this code:

### Generic part

```elisp
(defun hydra-generic-timeout (timer secs fun)
  (when (and (boundp timer)
             (timerp (symbol-value timer)))
    (cancel-timer (symbol-value timer)))
  (setq timer (set timer (timer-create)))
  (timer-set-time timer
                  (timer-relative-time (current-time) secs))
  (timer-set-function timer fun)
  (timer-activate timer))
```

### Specific part 

Press <kbd>C-v</kbd> in quick succession to scroll. If you give it 1 second of rest, `dired-jump` will be called and the hydra will exit.

```elisp
(defun hydra-cv-reset-timer ()
  (hydra-generic-timeout
   'hydra-cv-timer
   1.0 (lambda ()
         (dired-jump)
         (setq hydra-deactivate t)
         (hydra-keyboard-quit))))

(defhydra hydra-cv (:pre hydra-cv-reset-timer)
  ("C-v" scroll-up-command))

(global-set-key (kbd "C-v") 'hydra-cv/body)
```

**Comment by sandric** [comment]
Yes! Thanks a lot @abo-abo . Although not completely, in this example example `dired-jump` called everytime in the end - no matter called second `C-v` or not. As I understood this could be done with simple `:post` function handler as in my example above. To be answer to my question - if pressed second time run function B instead of A - you should add some kind of global validator. Here's my working example I tweaked to follow this logic:

```elisp
          (defun hydra-generic-timeout (timer secs fun)
            (when (and (boundp timer)
                       (timerp (symbol-value timer)))
              (cancel-timer (symbol-value timer)))
            (setq timer (set timer (timer-create)))
            (timer-set-time timer
                            (timer-relative-time (current-time) secs))
            (timer-set-function timer fun)
            (timer-activate timer))

          (setq fzf-validator nil)

          (defun hydra-timed-fzf ()
            (hydra-generic-timeout
             'hydra-cv-timer
             1.0
             (lambda ()
               (unless fzf-validator
                 (message "fzf called"))
               (setq hydra-deactivate t)
               (hydra-keyboard-quit))))

          (defhydra hydra-fzf (:pre hydra-timed-fzf)
            ("C-H-w" (progn (setq fzf-validator t) (message "fzf-directory called"))))

          (global-set-key (kbd "C-H-w") 'hydra-fzf/body)
```
Its a little bit many words but now I can make what I want with this `hydra-cv-timer`, thanks again.
